src/joystick/n3ds/SDL_sysjoystick.c

```diff
@@ -0,0 +1,367 @@
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "../../SDL_internal.h"
+
+#ifdef SDL_JOYSTICK_N3DS
+
+/* This is the N3DS implementation of the SDL joystick API */
+
+#include <3ds.h>
+
+#include "../SDL_sysjoystick.h"
+#include "SDL_events.h"
+
+#define NB_BUTTONS 23
+
+/*
+  N3DS sticks values are roughly within +/-160
+  which is too small to pass the jitter tolerance.
+  This correction factor is applied to axis values
+  so they fit better in SDL's value range.
+*/
+#define CORRECTION_FACTOR_X SDL_JOYSTICK_AXIS_MAX / 160
+
+/*
+  The Y axis needs to be flipped because SDL's "up"
+  is reversed compared to libctru's "up"
+*/
+#define CORRECTION_FACTOR_Y -CORRECTION_FACTOR_X
+
+/*
+  Factors used to convert touchscreen coordinates to
+  SDL's 0-1 values. Note that the N3DS's screen is
+  internally in a portrait disposition so the
+  GSP_SCREEN constants are flipped.
+*/
+#define TOUCHPAD_SCALE_X 1.0f / GSP_SCREEN_HEIGHT_BOTTOM
+#define TOUCHPAD_SCALE_Y 1.0f / GSP_SCREEN_WIDTH
+
+typedef struct N3DSJoystickState
+{
+    u32 kDown;
+    u32 kUp;
+    circlePosition circlePos;
+    circlePosition cStickPos;
+    accelVector acceleration;
+    angularRate rate;
+} N3DSJoystickState;
+
+SDL_FORCE_INLINE void UpdateAxis(SDL_Joystick *joystick, N3DSJoystickState *previous_state);
+SDL_FORCE_INLINE void UpdateButtons(SDL_Joystick *joystick, N3DSJoystickState *previous_state);
+SDL_FORCE_INLINE void UpdateSensors(SDL_Joystick *joystick, N3DSJoystickState *previous_state);
+SDL_FORCE_INLINE void UpdateTouch(SDL_Joystick *joystick);
+
+static N3DSJoystickState current_state;
+static SDL_bool sensors_enabled = SDL_FALSE;
+
+static int
+N3DS_JoystickInit(void)
+{
+    hidInit();
+    HIDUSER_EnableAccelerometer();
+    HIDUSER_EnableGyroscope();
+    return 0;
+}
+
+static const char *
+N3DS_JoystickGetDeviceName(int device_index)
+{
+    return "Nintendo 3DS";
+}
+
+static int
+N3DS_JoystickGetCount(void)
+{
+    return 1;
+}
+
+static SDL_JoystickGUID
+N3DS_JoystickGetDeviceGUID(int device_index)
+{
+    /* GUID corresponds to the name "Nintendo 3DS map" */
+    SDL_JoystickGUID guid = { { 0x4e, 0x69, 0x6e, 0x74, 0x65, 0x6e, 0x64, 0x6f, 0x20, 0x33, 0x44, 0x53, 0x20, 0x6d, 0x61, 0x70 } };
```

Comment:
I'd suggest using `SDL_CreateJoystickGUIDForName()` here for consistency with other drivers.